### PR TITLE
[Feature](Nereids) Main framework for selecting rollup index.

### PR DESCRIPTION
**Pull request by wangshuo128**
# Proposed changes
First step of #12303 

## Problem summary

This is the first step for supporting rollup index selection for aggregate/unique key OLAP table.

This PR aims to select rollup index when the aggregate node is present and the aggregate function matches the value type. So pre-aggregation is turned on by default.  Cases that pre-aggregation should be turned off will be addressed in the next PR.

Main steps for rollup index selection: 

1. filter rollup indexes with all the required columns.
2. filter rollup indexes that match the key prefix most.
3. order the rollup indexes by row count, column count, rollup index id.

TODO remaining:
1. address cases that pre-aggregation should be turned off. (next PR)
2. add more test cases. 

Refactor
- Add `Project.getSlotToProducer` to extract a map from the project output slot to its producing expression.
- Add `Filter.getConjuncts` to split the filter condition to conjunctive predicates.
- Move the usage of `ExpressionReplacer` to `ExpressionUtils.replace(expr, replaceMap)` to simplify the code.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
5. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
6. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
7. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
8. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Kikyou1997** [comment]
ExpressionUtils already has a function to extract conjuncts from an expression.

**Comment by wangshuo128** [comment]
> ExpressionUtils already has a function to extract conjuncts from an expression.

Actually, `Filter. getConjuncts` would delegate to  `ExpressionUtils.extractConjunction`. Adding this method is to simplify the code.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollup.java [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Split conjuncts into equal-to and non-equal-to.
+    ///////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Split conjuncts input equal-to and non-equal-to.
+     */
+    private Map<Boolean, Set<String>> split(List<Expression> conjunct, Map<ExprId, String> exprIdToColName) {
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollup.java:288]: Use Another concreter method name?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollup.java [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+            OlapTable table,
+            MaterializedIndex rollup,
+            Set<String> equalColNames,
+            Set<String> nonEqualColNames) {
+        int matchCount = 0;
+        for (Column column : table.getSchemaByIndexId(rollup.getId())) {
+            if (equalColNames.contains(column.getName())) {
+                matchCount++;
+            } else if (nonEqualColNames.contains(column.getName())) {
+                // Unequivalence predicate's columns can match only first column in rollup.
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollup.java:271]: I know this logic is same as the origin selector, but would you like to explain why `Unequivalence predicate's columns can match only first column in rollup.`? thx

**Comment by Kikyou1997** [comment]
> > ExpressionUtils already has a function to extract conjuncts from an expression.
> 
> Actually, `Filter. getConjuncts` would delegate to `ExpressionUtils.extractConjunction`. Adding this method is to simplify the code.

ok

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollup.java [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+            this.type = result;
+        }
+
+        public static CheckResult createEqual(String name) {
+            return new CheckResult(name, ResultType.SUCCESS_EQUAL);
+        }
+
+        public static CheckResult createNonEqual(String name) {
+            return new CheckResult(name, ResultType.SUCCESS_NON_EQUAL);
+        }
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollup.java:322]: need a verb between `could` and `prefix` here

**Review by 924060929 — APPROVED** [review]

**Pull request merged**